### PR TITLE
Safekeeper peer recovery

### PR DESCRIPTION
**Pull request by arssher**

### CODE REVIEW

File: safekeeper/src/safekeeper.rs [hunk 1]
```diff
@@ -91,6 +91,59 @@ impl TermHistory {
...
     }
+
+    /// Find point of divergence between leader (walproposer) term history and
+    /// safekeeper. Arguments are not symmetrics as proposer history ends at
+    /// +infinity while safekeeper at flush_lsn.
+    /// C version is at walproposer SendProposerElected.
+    pub fn find_highest_common_point(
+        prop_th: &TermHistory,
+        sk_th: &TermHistory,
+        sk_wal_end: Lsn,
```
Comment (petuhovskiy) [inline safekeeper/src/safekeeper.rs:102]: Is`sk_wal_end` always >= than `sk_th.last().lsn`?

File: safekeeper/src/send_wal.rs [hunk 2]
```diff
@@ -680,7 +681,7 @@ impl<IO: AsyncRead + AsyncWrite + Unpin> ReplyReader<IO> {
     }
 }
 
-const POLL_STATE_TIMEOUT: Duration = Duration::from_secs(1);
```
[the remark this answers is not in the document]
Comment (arssher) [reply safekeeper/src/send_wal.rs:683]: Partially related, I fixed one test which failed due to this change:
https://github.com/neondatabase/neon/pull/5205/commits/2b91f507a8860b07555fedda5af493ab44dd2230
Pageserver likely relies on it, but 1s seems too low anyway.

File: safekeeper/src/safekeeper.rs (hunk 1 above)
Comment (arssher) [reply safekeeper/src/safekeeper.rs:102]: Yes, it's a reasonable assert to add.

Comment (arssher) [reply safekeeper/src/safekeeper.rs:102]: added assert

File: safekeeper/src/send_wal.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply safekeeper/src/send_wal.rs:683]: Hm, in some cases 5s can be too high... Anyway, we probably should adjust default configs in pageserver::walreceiver somewhere.

**Comment by petuhovskiy** [comment]
test_peer_recovery failed in CI (postgres 16):

- `test_peer_recovery`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-5205/6404640660/index.html#suites/82004ab4e3720b47bf78f312dabe7c55/2f2e4fd81fd335d5)

File: safekeeper/src/bin/safekeeper.rs [hunk 3]
```diff
@@ -105,6 +105,9 @@ struct Args {
     /// it during this period passed as a human readable duration.
     #[arg(long, value_parser= humantime::parse_duration, default_value = DEFAULT_HEARTBEAT_TIMEOUT, verbatim_doc_comment)]
     heartbeat_timeout: Duration,
+    /// Disable/enable peer recovery. Used for disabling it in tests.
+    #[arg(long, default_value = "true", action=ArgAction::Set)]
```
[the remark this answers is not in the document]
Comment (arssher) [reply safekeeper/src/bin/safekeeper.rs:109]: done

File: safekeeper/src/receive_wal.rs [hunk 4]
```diff
@@ -96,6 +99,18 @@ impl WalReceivers {
...
 
+    /// Get number of streaming walreceivers (normally 0 or 1) from compute.
+    pub fn get_num_streaming(self: &Arc<WalReceivers>) -> usize {
+        self.mutex
+            .lock()
+            .slots
+            .iter()
+            .flatten()
+            // conn_id.is_none skips recovery which also registers here
+            .filter(|s| s.conn_id.is_none() && matches!(s.status, WalReceiverStatus::Streaming))
```
[the remark this answers is not in the document]
Comment (arssher) [reply safekeeper/src/receive_wal.rs:110]: Good catch, true.

**Review by petuhovskiy — APPROVED** [review]

**Pull request merged**